pkg/puller/watcher.go

```diff
@@ -0,0 +1,184 @@
+}
+
+type ModelWrapper struct {
+	ModelSpec *v1beta1.ModelSpec
+	Time      time.Time
+	Stale     bool
+	Success   bool
+}
+
+func WatchConfig(configDir string, numWorkers int) {
+	log.Println("Entering watch")
+	w.configDir = configDir
+	w.numWorkers = numWorkers
+	w.WatchConfig()
+}
+
+func (w *Watcher) WatchConfig() {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer watcher.Close()
+
+	done := make(chan bool)
+	go func() {
+		for {
+			select {
+			case event, ok := <-watcher.Events:
+				if !ok {
+					return
+				}
+				isCreate := event.Op&fsnotify.Create != 0
+				eventPath := filepath.Clean(event.Name)
+				isDataDir := filepath.Base(eventPath) == "..data"
+				// TODO: Should we use atomic integer or timestamp??
+				timeNow := time.Now()
+				if isDataDir && isCreate {
+					symlink, _ := filepath.EvalSymlinks(eventPath)
+					file, err := ioutil.ReadFile(filepath.Join(symlink, constants.ModelConfigFileName))
+					modelConfigs := make(modelconfig.ModelConfigs, 0)
+					if err != nil {
+						log.Println("Error in reading file", err)
+					}
+					err = json.Unmarshal([]byte(file), &modelConfigs)
+					if err != nil {
+						log.Println("unable to marshall for", event, "with error", err)
+					}
+					for _, modelConfig := range modelConfigs {
+						modelName := modelConfig.Name
+						modelSpec := modelConfig.Spec
+						log.Println("Name:", modelName, "Spec:", modelSpec)
+						oldModelInterface, ok := w.ModelTracker.Load(modelName)
+						if !ok {
+							w.ModelTracker.Store(modelName, ModelWrapper{
+								ModelSpec: &modelSpec,
+								Time:      timeNow,
+								Stale:     true,
+								Success:   false,
+							})
+						} else {
+							oldModel := oldModelInterface.(ModelWrapper)
+							isSame := false
+							if oldModel.ModelSpec != nil {
+								isSame = cmp.Equal(*oldModel.ModelSpec, modelSpec)
+								log.Println("same", isSame, *oldModel.ModelSpec, modelSpec)
+							}
+							if isSame {
+								// Need to store new time, maybe worth to have seperate map?
+								w.ModelTracker.Store(modelName, ModelWrapper{
+									ModelSpec: &modelSpec,
+									Time:      timeNow,
+									Stale:     false,
+									Success:   false,
+								})
+							} else {
+								w.ModelTracker.Store(modelName, ModelWrapper{
+									ModelSpec: &modelSpec,
+									Time:      timeNow,
+									Stale:     true,
+									Success:   false,
+								})
+							}
+						}
+					}
+					// TODO: Maybe make parallel and more efficient?
+					w.ModelTracker.Range(func(key interface{}, value interface{}) bool {
+						modelName, modelWrapper := key.(string), value.(ModelWrapper)
+						if modelWrapper.Time.Before(timeNow) {
+							w.ModelTracker.Delete(modelName)
+							channel, ok := p.ChannelMap[modelName]
+							if !ok {
+								log.Println("Model", modelName, "was never added to channel map")
+							} else {
+								event := EventWrapper{
+									ModelSpec:      nil,
```

Comment:
Can we use more explicit action enum for deletion/adding/update?

Reply:
Do we care? I thought we only care about Load and Unload when it comes to a request